File: valdi_modules/share/ios/ShareHelper.m
#import "ShareHelper.h"
#import <UIKit/UIKit.h>

@implementation SCValdiShareHelper

+ (void)shareWithTitle:(NSString *)title
                  text:(NSString *)text
                   url:(NSString *)url {
  NSMutableArray *items = [NSMutableArray array];
  if (text.length > 0) {
    [items addObject:text];
  }
  if (url.length > 0) {
    NSURL *parsed = [NSURL URLWithString:url];
    if (parsed) {
      [items addObject:parsed];
    }
  }
  if (items.count == 0) {
    return;
  }
  dispatch_async(dispatch_get_main_queue(), ^{
    UIWindowScene *windowScene = nil;
    for (UIScene *scene in UIApplication.sharedApplication.connectedScenes) {
      if ([scene isKindOfClass:[UIWindowScene class]]) {
        UIWindowScene *ws = (UIWindowScene *)scene;
        if (ws.activationState == UISceneActivationStateForegroundActive) {
          windowScene = ws;
          break;
        }
      }
    }
    if (!windowScene) {
      return;
    }
    UIWindow *keyWindow = nil;
    for (UIWindow *w in windowScene.windows) {
      if (w.isKeyWindow) {
        keyWindow = w;
        break;
      }
    }
    UIViewController *root = keyWindow.rootViewController;
    if (!root) {
      return;
    }
    while (root.presentedViewController) {
      root = root.presentedViewController;
    }
    UIActivityViewController *activity = [[UIActivityViewController alloc] initWithActivityItems:items applicationActivities:nil];
    if (activity.popoverPresentationController) {
      activity.popoverPresentationController.sourceView = root.view;
      activity.popoverPresentationController.sourceRect = CGRectMake(CGRectGetMidX(root.view.bounds), CGRectGetMidY(root.view.bounds), 0, 0);
      activity.popoverPresentationController.permittedArrowDirections = 0;
    }
    [root presentViewController:activity animated:YES completion:nil];
  });
}

@end

File: valdi_modules/share/ios/ShareHelper.swift
import UIKit

/// Native iOS implementation for the polyglot Share module.
/// Presents UIActivityViewController for the system share sheet.
/// The host app should wire Device.share(options) to ShareHelper.share(...).
@objc(SCValdiShareHelper)
public final class ShareHelper: NSObject {

    /// Show the system share sheet with the given content.
    /// Call from the native bridge when Share.share(options) is invoked from TS.
    @objc public static func share(title: String?, text: String?, url: String?) {
        var items: [Any] = []
        if let t = text, !t.isEmpty { items.append(t) }
        if let u = url, !u.isEmpty, let parsed = URL(string: u) { items.append(parsed) }
        if items.isEmpty { return }
        DispatchQueue.main.async {
            guard let windowScene = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .first(where: { $0.activationState == .foregroundActive }),
                  let root = windowScene.windows.first(where: { $0.isKeyWindow })?.rootViewController
            else { return }
            var top = root
            while let presented = top.presentedViewController { top = presented }
            let activity = UIActivityViewController(activityItems: items, applicationActivities: nil)
            if let popover = activity.popoverPresentationController {
                popover.sourceView = top.view
                popover.sourceRect = CGRect(x: top.view.bounds.midX, y: top.view.bounds.midY, width: 0, height: 0)
                popover.permittedArrowDirections = []
            }
            top.present(activity, animated: true)
        }
    }
}
